test/integration.ts

```diff
@@ -0,0 +1,236 @@
+    STREAMR_API_URL,
+    CoreAPIClient.DEFAULT_FETCH_FUNC,
+    CoreAPIClient.DEFAULT_GET_SESSION_TOKEN_FUNC,
+    adminKey
+)
+const trustedApiClient = new CoreAPIClient(
+    STREAMR_API_URL,
+    CoreAPIClient.DEFAULT_FETCH_FUNC,
+    CoreAPIClient.DEFAULT_GET_SESSION_TOKEN_FUNC,
+    DEVOPS_KEY
+)
+
+/**
+ * Resolves the promise once stream contains a match for target regex
+ * @param {Readable} stream to subscribe to
+ * @param {RegExp} regex to use for matching
+ * @returns {Match} the regex match object
+ */
+async function untilStreamMatches(stream: NodeJS.ReadableStream, regex: RegExp) {
+    return new Promise((resolve) => {
+        function check(buffer: Buffer) {
+            const data = buffer.toString()
+            const match = data.match(regex)
+            if (match) {
+                if (stream.off) { stream.off("data", check) }
+                resolve(match)
+            }
+        }
+        stream.on("data", check)
+    })
+}
+
+describe("Watcher", () => {
+
+    let productId: string
+    let productIdBytes: string
+    before(async function () {
+        this.timeout(30000)
+
+        // fund the devops key
+        const fundTx1 = await wallet.sendTransaction({
+            to: watcherWallet.address,
+            value: parseEther("10")
+        })
+        await fundTx1.wait()
+        log("Watcher balance mainnet: %s", (await watcherWallet.getBalance()).toString())
+
+        const fundTx2 = await sidechainWallet.sendTransaction({
+            to: watcherWallet.address,
+            value: parseEther("10")
+        })
+        await fundTx2.wait()
+        log("Watcher balance sidechain: %s", (await sidechainProvider.getBalance(watcherWallet.address)).toString())
+
+        const streamTx = await registry.createStream(streamIdPath, "{}")
+        log("Creating stream %s", streamId)
+        const streamTr = await streamTx.wait()
+        log("Events: %o", streamTr.events)
+
+        // create product into the core-api (simulate marketplace)
+        // @ts-ignore-next-line private
+        const createRes = await ownerApiClient._post(`${STREAMR_API_URL}/products/`, {
+            type: "DATAUNION",
+            state: "NOT_DEPLOYED",
+            beneficiaryAddress: watcherWallet.address,
+            dataUnionVersion: 2
+        })
+        const createResJson = await createRes.json()
+        log("Created product in core-api: %o", createResJson)
+        productId = createResJson.id
+        productIdBytes = "0x" + productId
+
+        // add stream to product
+        const streamIdEncoded = encodeURIComponent(streamId)
+        let status = 0
+        do {
+            // @ts-ignore-next-line private
+            const addRes = await ownerApiClient._post(`${STREAMR_API_URL}/products/${productId}/streams/${streamIdEncoded}`, {}, "PUT")
+            log("Add stream to product returned %s %s", addRes.status, addRes.statusText)
+            status = addRes.status
+        } while (status !== 204)
```

Comment:
that was a hack to work around web3j timing out on the server. There is no reason it should return 4xx or 5xx here.

Reply:
What if 204 never happens?

Reply:
then the test will fail with timeout. It should happen every time, but doesn't because of that mystery timeout (that I'm totally not going to debug now :D )